### PR TITLE
Add a wrapper to transform a NumPy function into a PyTorch function

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114610

A less general version of this wrapper was used in the keynote on
`torch.compile(numpy)`. We expose a generic version of the wrapper
that works seamlessly with `torch.compile`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: docs/source/torch.compiler_faq.rst [hunk 1]
```diff
@@ -614,23 +614,58 @@ to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors:
...
+on CPU. Note also that calling ``.numpy(force=True`` under ``torch.compile`` also
+moves to CPU and deactivates gradient computations.
+
+We provide the decorator ``torch.compiler.wrap_np``  that implements the patteron of wrapping
+a NumPy function into a PyTorch one. Using it, we can simply write
+
+.. code-block:: python
+
+   @torch.compile
+   @torch.compiler.wrap_fn
```
Comment (jansel) [inline docs/source/torch.compiler_faq.rst:642]: Should this (and below) be `wrap_np`?  Doesn't seem to match code.

File: torch/_dynamo/external_utils.py [hunk 2]
```diff
@@ -27,3 +35,21 @@ def call_hook(hook, *args):
     if result is None:
         return args[0]
     return result
+
+
+def wrap_np(f):
```
Comment (jansel) [inline torch/_dynamo/external_utils.py:40]: Should we call this wrap_numpy?  Or something more descriptive?  Wrap is pretty generic...

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/external_utils.py (hunk 2 above)
Comment (lezcano) [reply torch/_dynamo/external_utils.py]: Renamed to the more explicit `wrap_numpy`

**Review by jansel — COMMENTED** [review]
Code changes look good to me, though I'd like others to weigh in on if this should be a public API and what it should be called.

**Comment by albanD** [comment]
I'm not sure the semantic of this transformation is very clear to me and when I should use it?

Maybe adding type annotation to the functions will make it easier to see what happens?

**Comment by lezcano** [comment]
The type annotation would be `Callable[Any, Any] -> Callable[Any, Any]`, so I don't think it would add all that much...

**Comment by albanD** [comment]
I mean in terms of Tensor vs array.

**Comment by lezcano** [comment]
Sorry, I completely dropped the ball on this one. Added type annotations to all functions. It's not 100% clear to me whether this makes things simpler or not, given that the decorator changes the types, but yeah...

File: docs/source/torch.compiler_faq.rst [hunk 3]
```diff
@@ -614,23 +614,58 @@ to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors:
...
 them there. In this case ``X.numpy()`` and ``from_numpy()`` are hints to the compiler
 but no real data movement happens. Note that the original program would not run
 on eager mode now. If you want to run it in eager mode, you would need to call
 ``.numpy(force=True)`` doing ``Z = Z.cuda()`` before returning
 ``Z``. Of course, doing this would execute the program on eager mode NumPy, and
-on CPU.
+on CPU. Note also that calling ``.numpy(force=True`` under ``torch.compile`` also
+moves to CPU and deactivates gradient computations.
+
+We provide the decorator ``torch.compiler.wrap_numpy``  that implements the patteron of wrapping
```
Comment (albanD) [inline docs/source/torch.compiler_faq.rst:636]: I would move this at the top actually. Users should use this instead of doing it by hand in most cases?

File: torch/compiler/__init__.py [hunk 4]
```diff
@@ -118,3 +119,31 @@ def rand_foo():
...
     cudagraph_trees.mark_step_begin()
+
+def wrap_numpy(fn):
+    r"""Decorator that turns a function from ``np.ndarray``s to ``np.ndarray``s into a function
+    from ``torch.Tensor``s to ``torch.Tensor``s.
+
+    It is designed to be used with :func:`torch.compile`. It allows to compile a NumPy function
+    as if it were a PyTorch function. If the function can be traced without graph breaks,
+    this allows you to run NumPy code on CUDA.
+    Similarly, you can pass Tensors with ``requires_grad=True`` and then call ``backward()`` on the output
```
Comment (albanD) [inline torch/compiler/__init__.py:130]: nit to shorten the doc, this can be folded in the sentence above: `code on CUDA and with gradients enabled.`

File: docs/source/torch.compiler_faq.rst [hunk 5]
```diff
@@ -614,23 +614,58 @@ to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors:
...
+.. code-block:: python
+
+   @torch.compile
+   @torch.compiler.wrap_numpy
+   def numpy_fn(X: np.ndarray, Y: np.ndarray):
+       return np.sum(X[:, :, None] * Y[:, None, :], axis=(-2, -1))
+
+   X = torch.randn(1024, 64, device="cuda", requires_grad=True)
+   Y = torch.randn(1024, 64, device="cuda")
+   with torch.device("cuda"):
```
Comment (albanD) [inline docs/source/torch.compiler_faq.rst:664]: The context manager is not needed here actually? If so I think we should remove it from here.

**Review by albanD — COMMENTED** [review]
I think that there is a pretty big constraint here that we end up with a function that can ONLY be used in the context of compile-full-graph and will NOT work in eager.
This is a new enough concept that I think we should say it explicitly in the faq and docstring.

File: docs/source/torch.compiler_faq.rst (hunk 5 above)
Comment (lezcano) [reply docs/source/torch.compiler_faq.rst:664]: The context manager allows for graphbreaks to happen in the compiled code. Let me discuss graph breaks in the FAQ as well.

**Comment by lezcano** [comment]
Addressed the review. In particular:
- I have streamlined the examples in the FAQ, using only `wrap_numpy` as it's the preferred approach 
- I have added a paragraph discussing graph breaks explicitly.
- I have added a note discussing graph breaks explicitly to the docs.
  - Note that the docs are not currently rendered, as the `torch.compiler` page of the docs is currently being used as a page linking to other pages.

File: docs/source/torch.compiler_faq.rst [hunk 6]
```diff
@@ -602,35 +602,84 @@ context. Consider the example
...
+   assert isinstance(Z, np.ndarray)
 
 In this example, ``numpy_fn`` will be executed in CUDA. For this to be
 possible, ``torch.compile`` automatically moves ``X`` and ``Y`` from CPU
 to CUDA, and then it moves the result ``Z`` from CUDA to CPU. If we are
 executing this function several times in the same program run, we may want
 to avoid all these rather expensive memory copies. To do so, we just need
-to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors:
+to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors.
+We can do so by using ``torch.compiler.wrap_numpy``:
```
Comment (albanD) [inline docs/source/torch.compiler_faq.rst:613]: Thanks for the update, it looks quite good.
Should this piece be split into two questions in the FAQ?
- How to speed up NumPy code with `torch.compile`?
  - Use the fast cpu backend by enabling torch.compile
  - Run things on CUDA
- How to use NumPy code within your PyTorch code?
  - The wrap_numpy function

Comment (lezcano) [reply docs/source/torch.compiler_faq.rst:613]: I think splitting it is too much given how little information there is. I will add one example where we simply run things on CPU, implementing your first point "Use the fast cpu backend by enabling torch.compile"

Comment (lezcano) [reply docs/source/torch.compiler_faq.rst:642]: Nevermind, after writing it I chose to split it into two sections, although they are different to the ones you proposed.

**Review by albanD — APPROVED** [review]
Thanks for the update!
